File: Python/LLRBTree/Tree.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

"""
Tree.py: Based on Sedgewick's "Left-leaning Red-Black Tree".

Red-black trees are binary search trees obeying two key invariants:
1) All paths from root to leaf have same number of black nodes.
2) Red nodes always have black children.
"""

__copyright__ = "(c) 2016 E. Adrián Garro S. Costa Rica Institute of Technology."


class Tree:
    def __init__(self):
        self.red = True
        self.black = False
        # root of the Tree
        self.root = None

    # tree helper node data type
    class Node:
        def __init__(self, key, value, color):
            # key
            self.key = key
            # associated data
            self.value = value
            # color of parent link
            self.color = color
            # link to left subtree
            self.left = None
            # link to right subtree
            self.right = None

    def is_red(self, x):
        """
        Check if color of node x is red.
        """
        if not x:
            return False
        return x.color == self.red

    def is_empty(self):
        """
        Is this tree empty?
        """
        return self.root is None

    def search(self, key):
        """
        Returns the value associated with the given key.
        raise IndexError if the search returns None.
        """
        result = self.search_aux(self.root, key)
        if not result:
            raise IndexError("Key not found.")
        return result

    @staticmethod
    def search_aux(x, key):
        """
        Value associated with the given key in subtree rooted at x;
        None if no such key.
        """
        while x and key != x.key:
            if key < x.key:
                x = x.left
            else:
                x = x.right
        if x:
            return x.value

    def contains(self, key):
        """
        Does this tree contain the given key?
        """
        return self.search_aux(self.root, key) is not None

    def insert(self, key, value):
        """
        Inserts the specified key-value pair into the tree,
        overwriting the old value with the new value if the
        tree already contains the specified key.
        """
        self.root = self.insert_aux(self.root, key, value)
        self.root.color = self.black

    def insert_aux(self, h, key, value):
        """
        Insert the key-value pair in the subtree rooted at h.
        """
        if not h:
            return self.Node(key, value, self.red)
        if key < h.key:
            h.left = self.insert_aux(h.left, key, value)
        elif key > h.key:
            h.right = self.insert_aux(h.right, key, value)
        else:
            h.value = value
        # fix-up any right-leaning links
        if self.is_red(h.right) and not self.is_red(h.left):
            h = self.rotate_left(h)
        if self.is_red(h.left) and self.is_red(h.left.left):
            h = self.rotate_right(h)
        if self.is_red(h.left) and self.is_red(h.right):
            self.flip_colors(h)
        return h

    def delete_min(self):
        """
        Removes the smallest key and associated value from the tree.
        raise RuntimeWarning if the tree is empty.
        """
        if self.is_empty():
            raise RuntimeWarning("BST underflow.")
        # if both children of root are black, set root to red
        if not self.is_red(self.root.left) and not self.is_red(self.root.right):
            self.root.color = self.red
        self.root = self.delete_min_aux(self.root)
        if not self.is_empty():
            self.root.color = self.black

    def delete_min_aux(self, h):
        """
        Delete the key-value pair with the minimum key rooted at h.
        """
        if h.left:
            if not self.is_red(h.left) and not self.is_red(h.left.left):
                h = self.move_red_left(h)
            h.left = self.delete_min_aux(h.left)
            return self.balance(h)

    def delete(self, key):
        """
        Removes the specified key and its associated value
        from this tree (if the key is in this tree).
        raise indexError if key is not in tree.
        """
        if not self.contains(key):
            raise IndexError("Key not found.")
        # if both children of root are black, set root to red
        if not self.is_red(self.root.left) and not self.is_red(self.root.right):
            self.root.color = self.red
        self.root = self.delete_aux(self.root, key)
        if not self.is_empty():
            self.root.color = self.black

    def delete_aux(self, h, key):
        """
        Delete the key-value pair with the given key rooted at h.
        """
        if key < h.key:
            if not self.is_red(h.left) and not self.is_red(h.left.left):
                h = self.move_red_left(h)
            h.left = self.delete_aux(h.left, key)
        else:
            if self.is_red(h.left):
                h = self.rotate_right(h)
            if key == h.key and not h.right:
                return None
            if not self.is_red(h.right) and not self.is_red(h.right.left):
                h = self.move_red_right(h)
            if key == h.key:
                x = self.min_aux(h.right)
                h.key = x.key
                h.value = x.value
                h.right = self.delete_min_aux(h.right)
            else:
                h.right = self.delete_aux(h.right, key)
        return self.balance(h)

    def clear(self):
        """
        Delete all nodes from tree.
        """
        self.clear_aux(self.root)
        self.root = None

    def clear_aux(self, x):
        if x:
            # recurse: visit all nodes in the two subtrees.
            self.clear_aux(x.left)
            self.clear_aux(x.right)
            # after both subtrees have been visited,
            # set links of this node to None
            x.left = None
            x.right = None

    def rotate_right(self, h):
        """
        Make a left-leaning link lean to the right.
        """
        x = h.left
        h.left = x.right
        x.right = h
        x.color = x.right.color
        x.right.color = self.red
        return x

    def rotate_left(self, h):
        """
        Make a right-leaning link lean to the left
        """
        x = h.right
        h.right = x.left
        x.left = h
        x.color = x.left.color
        x.left.color = self.red
        return x

    @staticmethod
    def flip_colors(h):
        """
        Flip the colors of a node and its two children;
        h must have opposite color of its two children.
        """
        h.color = not h.color
        h.left.color = not h.left.color
        h.right.color = not h.right.color

    def move_red_left(self, h):
        """
        Assuming that h is red and both h.left and h.left.left
        are black, make h.left or one of its children red.
        """
        self.flip_colors(h)
        if self.is_red(h.right.left):
            h.right = self.rotate_right(h.right)
            h = self.rotate_left(h)
            self.flip_colors(h)
        return h

    def move_red_right(self, h):
        """
        Assuming that h is red and both h.right and h.right.left
        are black, make h.right or one of its children red.
        """
        self.flip_colors(h)
        if self.is_red(h.left.left):
            h = self.rotate_right(h)
            self.flip_colors(h)
        return h

    def balance(self, h):
        """
        Restore red-black tree invariant.
        """
        if self.is_red(h.right):
            h = self.rotate_left(h)
        if self.is_red(h.left) and self.is_red(h.left.left):
            h = self.rotate_right(h)
        if self.is_red(h.left) and self.is_red(h.right):
            self.flip_colors(h)
        return h

    def min(self):
        """
        Returns the smallest key in the tree,
        raise RuntimeWarning if the tree is empty.
        """
        if self.is_empty():
            raise RuntimeWarning("Called min with empty tree.")
        return self.min_aux(self.root).key

    @staticmethod
    def min_aux(x):
        """
        Node with smallest key in subtree rooted at x.
        """
        while x.left:
            x = x.left
        return x

    def max(self):
        """
        Returns the biggest key in the tree,
        raise RuntimeWarning if the tree is empty.
        """
        if self.is_empty():
            raise RuntimeWarning("Called max with empty tree.")
        return self.max_aux(self.root).key

    @staticmethod
    def max_aux(x):
        """
        Node with biggest key in subtree rooted at x;
        """
        while x.right:
            x = x.right
        return x

    def inorder_walk(self):
        return self.inorder_walk_aux(self.root)

    @staticmethod
    def inorder_walk_aux(x):
        s = list()
        while s or x:
            if x:
                s.append(x)
                x = x.left
            else:
                x = s.pop()
                yield x.value
                x = x.right

    def preorder_walk(self):
        return self.preorder_walk_aux(self.root)

    @staticmethod
    def preorder_walk_aux(x):
        if x:
            depth = 0
            s = list()
            s.append((x, depth))
            while s:
                node_and_depth = s.pop()
                x = node_and_depth[0]
                depth = node_and_depth[1]
                yield node_and_depth
                if x.right:
                    s.append((x.right, depth+1))
                if x.left:
                    s.append((x.left, depth+1))

    def pretty_print(self):
        for node, depth in self.preorder_walk():
            print(depth * "  ", end="")
            print("+-", end="")
            if node.color == self.red:
                print("\033[91m" + str(node.value) + "\033[0m")
            else:
                print("\033[90m" + str(node.value) + "\033[0m")

    def is_bst(self):
        """
        Returns true if a binary tree is a binary search tree.
        """
        if not self.is_empty():
            min_key = self.min()
            max_key = self.max()
            return self.is_bst_aux(self.root, min_key, max_key)
        return True

    def is_bst_aux(self, x, min_key, max_key):
        """
        Returns true if a binary tree is a binary search tree.
        """
        if not x:
            return True
        if x.key < min_key:
            return False
        if x.key > max_key:
            return False
        return (
            self.is_bst_aux(x.left, min_key, x.key)
            and self.is_bst_aux(x.right, x.key, max_key)
        )

    def is_balanced(self):
        """
        Do all paths from x to leaf have same number of black nodes?
        """
        # number of black links on path from root to min.
        black_links = 0
        x = self.root
        while x:
            if not self.is_red(x):
                black_links += 1
            x = x.left
        return self.is_balanced_aux(self.root, black_links)

    def is_balanced_aux(self, x, black_links):
        """
        Do all paths from x to leaf have same number of black nodes?
        """
        if not x:
            return black_links == 0
        if not self.is_red(x):
            black_links -= 1
        return (
            self.is_balanced_aux(x.left, black_links)
            and self.is_balanced_aux(x.right, black_links)
        )

File: Julia/LLRB/tree.jl
# Tree.jl: Based on Sedgewick's "Left-leaning Red-Black Tree".

# Red-black trees are binary search trees obeying two key invariants:
# 1) All paths from root to leaf have same number of black nodes.
# 2) Red nodes always have black children.

# (c) 2016 E. Adrián Garro S. Costa Rica Institute of Technology.

import Base.isempty

RED = true
BLACK = false

"""
Tree helper node data type.
"""
type Node{Key, Value}
	# key
	key::Key
	# associated data
	value::Value
	# color of parent link
	color::Bool
	# link to left subtree
	left::Nullable{Node{Key, Value}}
	# link to right subtree
	right::Nullable{Node{Key, Value}}
	# constructor
	function Node(key::Key, value::Value, color::Bool) 
		new(
			key, value, color, 
			Nullable{Node{Key, Value}}(),
			Nullable{Node{Key, Value}}()
		)
	end
end

"""
Tree structure.
"""
type Tree{Key, Value}
	# root of the Tree
	root::Nullable{Node{Key, Value}}
	function Tree()
	  new(Nullable{Node{Key, Value}}())
	end
end

"""
Check if color of node x is red.
"""
function isred{Key, Value}(x::Nullable{Node{Key, Value}})
	if isnull(x)
		false
	else
		get(x).color == RED
	end
end

"""
Is tree t empty?
"""
function isempty{Key, Value}(t::Tree{Key, Value})
	isnull(t.root)
end

"""
Returns the value associated with the given key.
return error if the search returns a null node.
"""
function search{Key, Value}(t::Tree{Key, Value}, key::Key)
	result = search(t.root, key)
	if isnull(result)
		error("Key not found.")
	end
	get(result)
end


"""
Value associated with the given key in subtree rooted at x;
null value if no such key.
"""
function search{Key, Value}(x::Nullable{Node{Key, Value}}, key::Key)
	while !isnull(x) && key != get(x).key
		if key < get(x).key
			x = get(x).left
		else
			x = get(x).right
		end
	end		
	if isnull(x)
		return Nullable{Value}()
	else
		return Nullable(get(x).value)
	end
end


"""
Does this tree contain the given key?
"""
function contains{Key, Value}(t::Tree{Key, Value}, key::Key)
	!isnull(search(t.root, key))
end


"""
Inserts the specified key-value pair into the tree,
overwriting the old value with the new value if the
tree already contains the specified key.
"""
function insert!{Key, Value}(t::Tree{Key, Value}, key::Key, value::Value)
	t.root = insert!(t.root, key, value)
	get(t.root).color = BLACK
end


"""
Insert the key-value pair in the subtree rooted at h.
"""
function insert!{Key, Value}(h::Nullable{Node{Key, Value}}, key::Key, value::Value)
	if isnull(h)
		Nullable(Node{Key, Value}(key, value, RED))
	else
		if key < get(h).key
			get(h).left = insert!(get(h).left, key, value)
		elseif key > get(h).key
			get(h).right = insert!(get(h).right, key, value)
		else
			get(h).value = value
		end
		# fix-up any right-leaning links
		if isred(get(h).right) && !isred(get(h).left)
			h = rotate_left(h)
		end	
		if isred(get(h).left) && isred(get(get(h).left).left)
			h = rotate_right(h)
		end	
		if isred(get(h).left) && isred(get(h).right)
			flip_colors!(h)
		end	
		h	
	end	
end

"""
Removes the smallest key and associated value from the tree.
error if the tree is empty.
"""
function delete_min!{Key, Value}(t::Tree{Key, Value})
	if isempty(t)
		error("BST underflow.")
	end	
	# if both children of root are black, set root to red
	if !isred(get(t.root).left) && !isred(get(t.root).right)
		get(t.root).color = RED
	end	
	t.root = delete_min!(t.root)
	if !isempty(t)
		get(t.root).color = BLACK
	end	
end		
		
		
"""
delete! the key-value pair with the minimum key rooted at h.
"""
function delete_min!{Key, Value}(h::Nullable{Node{Key, Value}})
	if !isnull(get(h).left)
		if !isred(get(h).left) && !isred(get(get(h).left).left)
			h = move_red_left(h)
		end	
		get(h).left = delete_min!(get(h).left)
		balance(h)
	end	
end		

"""
Removes the specified key and its associated value
from this tree (if the key is in this tree).
error if key is not in tree.
"""
function delete!{Key, Value}(t::Tree{Key, Value}, key::Key)
	if !contains(t, key)
		error("Key not found.")
	end	
	# if both children of root are black, set root to red
	if !isred(get(t.root).left) && !isred(get(t.root).right)
		get(t.root).color = RED
	end	
	t.root = delete!(t.root, key)
	if !isempty(t)
		get(t.root).color = BLACK
	end	
end		


"""
delete! the key-value pair with the given key rooted at h.
"""
function delete!{Key, Value}(h::Nullable{Node{Key, Value}}, key::Key)
	if key < get(h).key
		if !isred(get(h).left) && !isred(get(get(h).left).left)
			h = move_red_left(h)
		end	
		get(h).left = delete!(get(h).left, key)
	else
		if isred(get(h).left)
			h = rotate_right(h)
		end	
		if key == get(h).key && isnull(get(h).right)
			return Nullable{Node{Key, Value}}()
		end	
		if !isred(get(h).right) && !isred(get(get(h).right).left)
			h = move_red_right(h)
		end	
		if key == get(h).key
			x = min(get(h).right)
			get(h).key = get(x).key
			get(h).value = get(x).value
			get(h).right = delete_min!(get(h).right)
		else
			get(h).right = delete!(get(h).right, key)
		end
	end		
	balance(h)
end	

"""
Delete all nodes from tree.
"""
function clear!{Key, Value}(t::Tree{Key, Value})
	clear!(t.root)
	t.root = Nullable{Node{Key, Value}}()
end	

function clear!{Key, Value}(x::Nullable{Node{Key, Value}})
	if !isnull(x)
		# recurse: visit all nodes in the two subtrees.
		clear!(get(x).left)
		clear!(get(x).right)
		# after both subtrees have been visited,
		# set links of this node to None
		get(x).left = Nullable{Node{Key, Value}}()
		get(x).right = Nullable{Node{Key, Value}}()
	end
end		

"""
Make a left-leaning link lean to the right.
"""
function rotate_right{Key, Value}(h::Nullable{Node{Key, Value}})
	x = get(h).left
	get(h).left = get(x).right
	get(x).right = h
	get(x).color = get(get(x).right).color
	get(get(x).right).color = RED
	x
end

"""
Make a right-leaning link lean to the left.
"""
function rotate_left{Key, Value}(h::Nullable{Node{Key, Value}})
	x = get(h).right
	get(h).right = get(x).left
	get(x).left = h
	get(x).color = get(get(x).left).color
	get(get(x).left).color = RED
	x
end

"""
Flip the colors of a node and its two children,
h must have opposite color of its two children.
"""
function flip_colors!{Key, Value}(h::Nullable{Node{Key, Value}})
	get(h).color = !get(h).color
	get(get(h).left).color = !get(get(h).left).color
	get(get(h).right).color = !get(get(h).right).color
end

"""
Assuming that h is red and both h.left and h.left.left
are black, make h.left or one of its children red.
"""
function move_red_left{Key, Value}(h::Nullable{Node{Key, Value}})
	flip_colors!(h)
	if isred(get(get(h).right).left)
		get(h).right = rotate_right(get(h).right)
		h = rotate_left(h)
		flip_colors!(h)
	end	
	h
end	
	
	
"""
Assuming that h is red and both h.right and h.right.left
are black, make h.right or one of its children red.
"""
function move_red_right{Key, Value}(h::Nullable{Node{Key, Value}})
	flip_colors!(h)
	if isred(get(get(h).left).left)
		h = rotate_right(h)
		flip_colors!(h)
	end	
	h
end	

"""
Restore red-black tree invariant.
"""
function balance{Key, Value}(h::Nullable{Node{Key, Value}})
	if isred(get(h).right)
		h = rotate_left(h)
	end	
	if isred(get(h).left) && isred(get(get(h).left).left)
		h = rotate_right(h)
	end	
	if isred(get(h).left) && isred(get(h).right)
		flip_colors!(h)
	end	
	h
end	

"""
Returns the smallest key in the tree,
error if the tree is empty.
"""
function min{Key, Value}(t::Tree{Key, Value})
	if isempty(t)
		error("Called min with empty tree.")
	end	
	get(min(t.root)).key
end	
	
	
"""
Node with smallest key in subtree rooted at x;
"""
function min{Key, Value}(x::Nullable{Node{Key, Value}})
	while !isnull(get(x).left)
		x = get(x).left
	end	
	x
end

"""
Returns the biggest key in the tree,
error if the tree is empty.
"""
function max{Key, Value}(t::Tree{Key, Value})
	if isempty(t)
		error("Called max with empty tree.")
	end	
	get(max(t.root)).key
end	
	
	
"""
Node with biggest key in subtree rooted at x;
"""
function max{Key, Value}(x::Nullable{Node{Key, Value}})
	while !isnull(get(x).right)
		x = get(x).right
	end	
	x
end
			
function inorderwalk{Key, Value}(t::Tree{Key, Value})
	inorderwalk(t.root)
end

function inorderwalk{Key, Value}(x::Nullable{Node{Key, Value}})
	s = []
	while !isempty(s) || !isnull(x)
		if !isnull(x)
			push!(s, x)
			x = get(x).left
		else
			x = pop!(s)
			produce(get(x).value)
			x = get(x).right
		end
	end	
end

function preorderwalk{Key, Value}(t::Tree{Key, Value})
	preorderwalk(t.root)
end

function preorderwalk{Key, Value}(x::Nullable{Node{Key, Value}})
	if !isnull(x)
		depth = 0
		s = []
		push!(s, (x, depth))
		while !isempty(s)
			node_and_depth = pop!(s)
			x = node_and_depth[1]
			depth = node_and_depth[2]
			produce(node_and_depth)
			if !isnull(get(x).right)
				push!(s, (get(x).right, depth + 1))
			end	
			if !isnull(get(x).left)
				push!(s, (get(x).left, depth + 1))
			end
		end
	end
end	

function prettyprint{Key, Value}(t::Tree{Key, Value})
	prettyprint(t.root, 0)
end

function prettyprint{Key, Value}(h::Nullable{Node{Key, Value}}, depth::Int64)
	for (node, depth) in @task preorderwalk(h)
		print("  " ^ depth)
		print("+-")
		if get(node).color == RED
			println(
				string("\033[91m", string(
						string(get(node).value), "\033[0m"
					)
				)
			)
		else
			println(
				string("\033[90m", string(
						string(get(node).value), "\033[0m"
					)
				)
			)
		end	
	end
end

"""
Returns true if a binary tree is a binary search tree.
"""
function isbst{Key, Value}(t::Tree{Key, Value})
	if !isempty(t)
		minkey = min(t)
		maxkey = max(t)
		isbst(t.root, minkey, maxkey)
	else
		true
	end
end

"""
Returns true if a binary tree is a binary search tree.
"""
function isbst{Key, Value}(x::Nullable{Node{Key, Value}}, minkey::Key, maxkey::Key)
	if isnull(x)
		return true
	end	
	if get(x).key < minkey
		return false
	end	
	if get(x).key > maxkey
		return false
	end	
	isbst(get(x).left, minkey, get(x).key) && isbst(get(x).right, get(x).key, maxkey)
end	

"""
Do all paths from x to leaf have same number of black nodes?
"""
function isbalanced{Key, Value}(t::Tree{Key, Value})
	# number of black links on path from root to min.
	blacklinks = 0
	x = t.root
	while !isnull(x)
		if !isred(x)
			blacklinks += 1
		end
		x = get(x).left
	end	
	isbalanced(t.root, blacklinks)
end

"""
Do all paths from x to leaf have same number of black nodes?
"""
function isbalanced{Key, Value}(x::Nullable{Node{Key, Value}}, blacklinks::Int64)
	if isnull(x)
		return blacklinks == 0
	end	
	if !isred(x)
		blacklinks -= 1
	end	
	isbalanced(get(x).left, blacklinks) && isbalanced(get(x).right, blacklinks)
end
